Shared/ApiShared.cs

```diff
@@ -86,11 +86,24 @@ public virtual object InternalCall(HttpMethod method, string url, SortedDictiona
             throw new Exception("Please call overriden method");
         }
 
+        internal virtual T CallApi<T>(HttpMethod method, string url, BaseParams parameters, FileDescription file, Dictionary<string, string> extraHeaders = null) where T : BaseResult, new()
+        {
+            SortedDictionary<string, object> parametersDict = null;
+            if (parameters != null)
+            {
+                parameters.Check();
+                parametersDict = parameters.ToParamsDictionary();
+            }
+
+            return CallAndParse<T>(method, url, parametersDict, file, extraHeaders);
```

Comment:
We are not supposed to use obsolete method by ourselves

Reply:
Should we just duplicate code of CallAndParse here or create another private not obsolete method with the same signature as CallAndParse? In case you'd like to just avoid CallAndParse without private replacement - there are some nuances with other uses:

- public Upload method with Dictionary as a parameter, so we can't use CallApi there. Should we mark it as obsolete as well? What should we add instead? 
- UploadPreset - doable, but we'll need to create a way to clone params object and modify serializing to a dictionary to ignore name if it's null
- Search Execute - currently has no params class, uses a dictionary. I'd suggest to keep it this way in current PR, as we discussed advanced search feature and we can make search params as part of it to do it once.
- UploadLarge & CallUploadMappingAPI - can be fixed.

Reply:
0. Remove obsolete attribute from CallAndParse, it will be still available for users who want to call invoke methods with arbitrary dict. Maybe in docstring we should write that it is not recommended.
1.  Leave Upload as it is.
2. UpdateUploadPreset - Should be fixed. We miss validation there, that's a bug!
You can create a shallow copy of params object, and set name to null.
Now, when values are empty, they are not added to the dictionary. see `AddParam` definition
3. Yes, that can be postponed for now.
4. Great!